### PR TITLE
test: Introduce Infection for mutation tests

### PR DESCRIPTION
**Pull request by Wirone**
After recent [discussion](https://twitter.com/maks_rafalko/status/1765841854702981331) I decided to verify what is our mutation score 😁. It does look good to be honest with initial setup:

```
23750 mutations were generated:
   18693 mutants were killed
       0 mutants were configured to be ignored
       0 mutants were not covered by tests
    4581 covered mutants were not detected
      70 errors were encountered
     136 syntax errors were encountered
     231 time outs were encountered
      39 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 80%
         Mutation Code Coverage: 100%
         Covered Code MSI: 80%
```

## Concerns to address

- Infection is currently the slowest part of our QA workflow (with _this initial_ configuration). We put a lot of work in the past to make our CI faster, so it's a no-go to add a job that takes much longer than the rest (21m30s on M1 with 10 cores). We need to determine if it's possible to improve it.
- There is a lot of `S` at the beginning, then it freezes for significant amount of time, then it runs further (why?).
- Mutating `src/Runner/Runner.php` causes some fixture files to be modified, which causes regular test failures and may lead to committing changes that should not be committed.

CC: @maks-rafalko, any suggestion are welcome 😃.

### CODE REVIEW

**Comment by maks-rafalko** [comment]
Hello, I'm glad you decided to try it :)

> * Infection is currently the slowest part of our QA workflow (with _this initial_ configuration). We put a lot of work in the past to make our CI faster, so it's a no-go to add a job that takes much longer than the rest (21m30s on M1 with 10 cores). We need to determine if it's possible to improve it.

you can start with another approach first: instead of mutating all the files and running all the tests, you can mutate only those *lines of code* that actually changed (added, updated) in the Pull Request.

Basically, we do exactly that on Infection itself: https://github.com/infection/infection/blob/873cd3335774a114bef9ca93388e623bf362d820/.github/workflows/mt-annotations.yaml#L55

Infection automatically shows github annotations on the lines that were mutated by mutant wasn't killed, so it's quite convenient ([example](https://github.com/infection/infection/pull/1909/files#diff-a26cf0383ec4e6a5f5bc01dbcec23ddf5931cb706fac3cf6b2eeabe9b53d0544R165)).

This is the fastest approach. Does it make sense?

> * There is a lot of `S` at the beginning, then it freezes for significant amount of time, then it runs further (why?).

Regarding `S`kipped mutants: please read https://infection.github.io/2020/08/18/whats-new-in-0.17.0/#Skip-S-mutations-that-are-over-specified-time-limit.

Regarding freeze: `S`kipped mutants are calculated very fast, that's why you see them first, and then others are calculated in different threads step by step, that's why you see some delay. This is ok from Infection point of view.

> * Mutating `src/Runner/Runner.php` causes some fixture files to be modified, which causes regular test failures and may lead to committing changes that should not be committed.

you can exclude any files from mutation process, please see here (`exclude` section in the config): https://infection.github.io/guide/usage.html

File: infection.json5.dist [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        "@default": true,
+        "Throw_": {
+            "ignore": [
+                "PhpCsFixer\\Cache\\FileHandler"
+            ]
+        }
+    },
+    "logs": {
+        "text": "dev-tools/infection/run.log",
+        "html": "dev-tools/infection/report.html"
```
Comment (maks-rafalko) [inline infection.json5.dist:19]: I think you don't need it on CI, neither `text` nor `html`

Comment (Wirone) [reply infection.json5.dist:19]: I wanted to create them for local runs, so they can be viewed and analysed. I will take a look how to do it better.

For now I need to figure out what's the problem in the CI, but I'll take a look later.

**Comment by keradus** [comment]
> Hello, I'm glad you decided to try it :)

I tried it out for Fixer already few years ago (2019-06), yet it was too many hours to run on CI back then. Eager to see how fast it can be on CI nowadays !

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -107,6 +107,11 @@ jobs:
             run-tests: 'yes'
             collect-code-coverage: 'yes'
 
+          - operating-system: 'ubuntu-20.04'
+            php-version: '8.3'
```
Comment (keradus) [inline .github/workflows/ci.yml:111]:
```suggestion
            php-version: ${{ env.PHP_MAX }}
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -197,6 +202,12 @@ jobs:
           FAST_LINT_TEST_CASES: 1
         run: vendor/bin/paraunit coverage --testsuite unit --pass-through=--exclude-group=covers-nothing --clover=build/logs/clover.xml
 
+      - name: Run mutation tests (Infection)
+        if: matrix.run-mutation-tests == 'yes'
+        env:
+          FAST_LINT_TEST_CASES: 1
+        run: composer infection
```
Comment (keradus) [inline .github/workflows/ci.yml:209]: *minor*

inconsistency: before we run `composer test` (and not `composer PHPUnit`) , but here we run `composer infection` (and not `composer mutation` etc)

**Comment by Wirone** [comment]
> you can start with another approach first: instead of mutating all the files and running all the tests, you can mutate only those lines of code that actually changed (added, updated) in the Pull Request. (...) This is the fastest approach. Does it make sense?

Of course, I saw it in the docs but wanted to start with full check anyway, just to check how the MSI looks like for our tests 🙂. But it seems like full check is a no-go, so I will take a look at the diff approach. My only doubt is how to approach MSI, because some files may have lower, some may have higher indicator. How should it be done properly? In your case you don't use `--min-covered-msi`, just `--git-diff-lines`, so I suppose it's the way to go (no minimal indicator, just find escaped mutants)?

> you can exclude any files from mutation process, please see here (exclude section in the config)

Yeah, I saw that. I think I'll try to figure out the exact mutator which causes this, I just did not have time for it yet. For now I was only able to ensure that `composer infection -- --filter=src/Runner/Runner.php` causes file modifications, so I'll try to narrow it down 🙂.

Thanks for great feedback @maks-rafalko, I'll improve this PR soon and if I have any more doubts, I'll let you know.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (Wirone) [reply .github/workflows/ci.yml:209]: I'll use `test:mutation` here (also available). `infection` is just a handy alias.

File: composer.json [hunk 4]
```diff
@@ -132,6 +135,10 @@
             "Composer\\Config::disableProcessTimeout",
             "paraunit run --testsuite integration"
         ],
+        "test:mutation": [
+            "Composer\\Config::disableProcessTimeout",
+            "infection --threads=max --only-covered --min-covered-msi=80"
```
Comment (dkarlovi) [inline composer.json:140]: You can use [`--only-covering-test-cases`](https://infection.github.io/guide/command-line-options.html#only-covering-test-cases) here which will run only test cases claiming to cover the line being mutated. It might make a difference for you assuming you have a lot of I/O in tests.

**Comment by maks-rafalko** [comment]
> My only doubt is how to approach MSI, because some files may have lower, some may have higher indicator. How should it be done properly? In your case you don't use `--min-covered-msi`, just `--git-diff-lines`, so I suppose it's the way to go (no minimal indicator, just find escaped mutants)?

Let me clarify it with more details just to be sure we are on the same page about all the aspects of Infection.

I think we have 2 goals here to use Infection:

1. to understand the current MSI (the quality of the tests) - it's done, MSI=80%. This is our baseline.
2. to _control_ the quality of the tests (MSI) and step by step improve it

**How can we control the quality (MSI)?**

 Only by forcing all the new code to be merged with MSI >= 80%. And from time to time increase this value (to 80.1%, then to 80.2% and so on).

**How can we force minimum MSI?**

Only by failing the build, thanks to `--min-msi` / `--min-covered-msi` (difference is described below). So from my understanding, `--min-covered-msi` this is a must have option here. And yes, it can be used together with `--git-diff-lines` and `--ignore-msi-with-no-mutations`.

**What value we should set `--min-covered-msi` to?** 

This depends on many factors. Good approach is to start with the baseline and continuously increase it as I described above, failing the builds if MSI < %baseline%.

Another approach is to set it to 100% and do not allow any `escaped` mutants. But this is quite hard for contributors and maintainers for such a big project - you will also have cases where old code is changed and developers will have to "fix" previous low-MSI code. And it's not that easy for new Infection users - you will have to understand each mutation, learn `ignore` features and be tolerant to the fact that there can be false-positive, which will take your time.

**Why do I suggest to start with at least `--git-diff-lines --min-covered-msi=80 --ignore-msi-with-no-mutations`?**

- `--git-diff-lines` is for performance - we will only mutate the diff (new and updated lines of code) instead of the whole project. Also, `--git-diff-lines` will show [inline annotations](https://github.com/infection/infection/pull/1909/files#diff-a26cf0383ec4e6a5f5bc01dbcec23ddf5931cb706fac3cf6b2eeabe9b53d0544R165) only for the changed code, which is nice with PR-based development.
- `--min-covered-msi=80` - new/updated code with MSI < 80% will not be allowed. It means that if we merge a piece of code with MSI >= 80%, then the whole project will have MSI >=80% and in fact will be continuously increasing.
- `--ignore-msi-with-no-mutations` - if we add/update lines of code that don't produce any mutations, covered MSI will be 0%. This option allows to not fail the build in this case. No mutations -> build is green.

----

I would also start with `--only-covered` option as you are probably interested in [covered MSI](https://infection.github.io/guide/#Covered-Code-Mutation-Score-Indicator), not just MSI.

In short, 

- MSI is to force developers write more tests, because uncovered lines of code will decrease MSI.
- Covered MSI is to focus on covered code and improve the quality of the tests that cover it, ignoring uncovered code.

**Comment by Wirone** [comment]
Thank you @maks-rafalko for the great explanation, I believe we're on the same page here as my vision was more or less as you described 😁. I am just a little afraid about the practice part 😅. In the end, the goal is to provide more tests and raise MSI, but as the baseline MSI (80) is only an average value, some contributions may hit the problem described by you - too low MSI just because of historically not-tested-enough code. In this project only @keradus have rights to merge PRs with failures, but really often in practice most of the reviews and merges are done by me, so without the ability to ignore MSI while merging I can see myself getting stuck with PRs 🤔. I could be forced to help people with achieving proper MSI, guide them or maybe even add tests / change Infection config by myself. It's doable, but far from ideal. From this perspective, the best initial introduction of mutation tests would be if we had inline hints for escaped mutants (to point people and maintainers what can/should be improved), but without failed CI. What do you think about it?

**Comment by maks-rafalko** [comment]
> From this perspective, the best initial introduction of mutation tests would be if we had inline hints for escaped mutants (to point people and maintainers what can/should be improved), but without failed CI. What do you think about it?

Sounds good. You can always add `--min-covered-msi` later if needed.

File: infection.json5.dist (hunk 1 above)
Comment (Wirone) [reply infection.json5.dist:19]: Since it's impossible to add these reports dynamically with CLI options, I'll leave it here because for diff-based testing in CI it shouldn't be an too big overhead.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (Wirone) [reply .github/workflows/ci.yml]: In the end I've used totally dedicated script for CI.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Wirone) [reply .github/workflows/ci.yml:111]: I had to revert this change as it was causing problem - none of CI jobs were added to the workflow. Maybe it should be in single quotes or without `${{ }}`, I don' know, but coverage job also is run only on 8.3 and has this version explicitly defined in the matrix. If we want to clean it up, we can do it later.

**Comment by Wirone** [comment]
@maks-rafalko can you help me with [this error](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8247458004/job/22555631934?pr=7874#step:11:36)? I don't get where this error comes from, as the dynamically generated config should be in sync with the schema provided with a tool 🤔.

**Comment by maks-rafalko** [comment]
I will be able to look into it not earlier that tonight, but I noticed that the old Infection version is used (`0.26.21`). We had many  fixes since then for PHPUnit, so please try to upgrade. If this is not possible due to locked conflicting versions, you can try using `infection.phar` (which is prefixed inside thanks to box+php-scoper)

**Comment by Wirone** [comment]
It's weird, because I added `^0.26.6` constraint to be able to use last version that supports PHP7.4, but at the same time to install latest version on 8.3. It seems like 0.27 is not installed neither in CI and for local development, but when I change to `^0.27` then it's installed 🤷‍♂️. I had to go out now, and will look more at it later.

**Comment by maks-rafalko** [comment]
This is expected from composer POV:

https://getcomposer.org/doc/articles/versions.md#caret-version-range-

> For pre-1.0 versions it also acts with safety in mind and treats `^0.3` as `>=0.3.0 <0.4.0` and `^0.0.3` as `>=0.0.3 <0.0.4`.

File: composer.json [hunk 5]
```diff
@@ -40,6 +40,7 @@
     },
     "require-dev": {
         "facile-it/paraunit": "^1.3 || ^2.0",
+        "infection/infection": "^0.26.6",
```
Comment (Wirone) [inline composer.json:43]:
```suggestion
        "infection/infection": "^0.27",
```

Since we run Infection on PHP8.3 in CI (and remove it completely on any other job), let's bump the minimal version. For local development it also will require 8.1, but we need it anyway because of `dev-tools`.

**Comment by Wirone** [comment]
Unfortunately there's [the same error](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8249729722/job/22562758513#step:11:35) on v0.27.10 😟. Currently on mobile, so can't dig much, will take a look later.

**Comment by maks-rafalko** [comment]
The current failure is because of the old PHPUnit xml configuration file format that needs to be migrated.

image

`vendor/bin/phpunit --migrate-configuration` will do the job.

After it, Infection will run (just tried locally).

If we look into other actions, there always different version of `phpunit/phpunit` is installed. E.g. [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8249729722/job/22562752218?pr=7874#step:4:204) it's `phpunit/phpunit (9.6.0)`.

But when we run Infection with PHPUnit on PHP 8.3, `phpunit/phpunit (10.5.12)` is installed, and the current xml file doesn't pass PHPUnit's schema validation, that Infection executes.

Since ``phpunit/phpunit ^9.6 || ^10.5.5 || ^11.0.2` is in the `composer.json` we can run `vendor/bin/phpunit --migrate-configuration` just before running Infection in GH Action.

Infection doesn't have an option to turn off validation, we do it intentionally to make sure project's PHPUnit configuration is in the correct format to make sure coverage (and all other features) work properly.

**Comment by Wirone** [comment]
Guys, it's working 😁 

image

In [this run](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8252639999) you can see warnings related to escaped mutants because I've made one indentation to a random file causing "change" of 1 whole class. I'll revert that change and I believe PR is ready 🥳.

Thank you @maks-rafalko for your assistance 🍻!

**Comment by maks-rafalko** [comment]
Looks good from my side. Great start!

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -151,12 +156,16 @@ jobs:
...
+        if: matrix.run-mutation-tests != 'yes'
+        run: composer remove --dev infection/infection --no-update
+
       - name: Setup PHP with Composer deps
         uses: ./.github/composite-actions/setup-php-with-composer-deps
         with:
           os: ${{ runner.os }}
           php: ${{ matrix.php-version }}
-          php-coverage: ${{ matrix.collect-code-coverage }}
+          php-coverage: ${{ matrix.collect-code-coverage || matrix.run-mutation-tests }}
```
Comment (keradus) [inline .github/workflows/ci.yml:168]: questio: do i understand right that we need code-coverage for running mutation tests?

Comment (maks-rafalko) [reply .github/workflows/ci.yml:168]: Yes. It can be generated by Infection (current usage) or passed as argument if pre-generated

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -235,6 +252,9 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v4
 
+      - name: Remove Infection
+        run: composer remove --dev infection/infection --no-update
```
Comment (keradus) [inline .github/workflows/ci.yml:256]: question: 
why do we need that? we do not remove manually any other dev deps for build job (and deps are cached anyway)

File: composer.json [hunk 8]
```diff
@@ -40,6 +40,7 @@
     },
     "require-dev": {
         "facile-it/paraunit": "^1.3 || ^2.0",
+        "infection/infection": "^0.27",
```
Comment (keradus) [inline composer.json:43]:
```suggestion
        "infection/infection": "~0.27",
```

due to semver, 0.28 can break BC breaks

File: composer.json (hunk 4 above)
Comment (keradus) [reply composer.json:139]: shall we add this before merge?

File: composer.json (hunk 8 above)
Comment (Wirone) [reply composer.json:43]: Again, [it's not needed](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7874#issuecomment-1991641902).

File: infection.json5.dist [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    "source": {
+        "directories": [
+            "src"
+        ]
+    },
+    "mutators": {
+        "@default": true,
+        "Throw_": {
+            "ignore": [
+                "PhpCsFixer\\Cache\\FileHandler"
```
Comment (keradus) [inline infection.json5.dist:13]: as it's json5 (nice! not many tools allow for that format!) we can add comment describing why we are ignoring this class

**Review by keradus — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 7 above)
Comment (Wirone) [reply .github/workflows/ci.yml:256]: Because if we don't do it, it will fail some jobs on 7.4 ☺️. Since we need it locally too, I could add it dynamically in the CI only in Infection job. We can't add it to dev-tools too, because it requires PHPUnit and it is installed in regular Composer, and we don't want 2 of them 🙂.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (keradus) [reply .github/workflows/ci.yml:168]: got you, thanks ;)

File: composer.json (hunk 8 above)
Comment (keradus) [reply composer.json:43]: TIL, nice

File: composer.json (hunk 4 above)
Comment (Wirone) [reply composer.json:139]: @keradus I would rather wait for infection/infection#1919 🙂.

Comment (maks-rafalko) [reply composer.json:139]: @Wirone released https://github.com/infection/infection/releases/tag/0.27.11

Docs: https://infection.github.io/guide/command-line-options.html#map-source-class-to-test

**Comment by Wirone** [comment]
@maks-rafalko [2m27s](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8252639999/job/22572631244?pr=7874#step:11:1) vs [8s](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8367119469/job/22908827843?pr=7874#step:12:1) 🥳!

image

image

Awesome improvement! 🍻

**Comment by maks-rafalko** [comment]
This is really awesome 😮🥳

File: infection.json5.dist (hunk 1 above)
Comment (Wirone) [reply infection.json5.dist]: In the end I removed them from dist config, everyone can add them to local config if needed 🙂.

File: infection.json5.dist (hunk 9 above)
Comment (Wirone) [reply infection.json5.dist:14]: Added explanation. I should ignore some mutants for `src/Runner/Runner.php` too, but did not find out which one yet. In terms of `Runner`, mutation tests "only" causes modifications of fixture files, but when it comes to `FileHandler` it was modifying file permissions in a way it required `sudo` to get it back (and unreadable file was causing failures in tests).

File: composer.json [hunk 10]
```diff
@@ -132,6 +134,10 @@
             "Composer\\Config::disableProcessTimeout",
             "paraunit run --testsuite integration"
         ],
+        "test:mutation": [
+            "Composer\\Config::disableProcessTimeout",
+            "infection --threads=max --only-covered --min-covered-msi=80"
```
Comment (julienfalque) [inline composer.json:139]: Shouldn't the CI command also use `--min-covered-msi=80`?

Comment (Wirone) [reply composer.json:139]: No, here we have average MSI calculated for all tests, but in PRs we run mutation tests only for modified classes and it may happen that they have lower MSI. In such cases, we only want to annotate code with escaped mutants, but we don't want to fail the job and force PR's author to improve MSI 🙂. For now it's strictly informative tool, that should help us with verifying quality of the tests in each PR.

On the other hand, `test:mutation` is for determining MSI in project, we can think about running it as a scheduled workflow at some point to observe the metric 🙂. But it's unlikely to be added in regular PRs.

Comment (bendavies) [inline composer.json:139]: `--only-covered --min-covered-msi` can go in `infection.json5.dist`. good to have config in one place?

Comment (Wirone) [reply composer.json:139]: `--min-covered-msi` can't be defined in the dist file because it's used both for Composer script (full analysis) and for CI, where [we don't require min MSI for PRs](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7874#discussion_r1533414437). `infection.json5.dist` must contain only truly common stuff, `--min-covered-msi` is contextual in our case.

**Comment by dkarlovi** [comment]
> @maks-rafalko [2m27s](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8252639999/job/22572631244?pr=7874#step:11:1) vs [8s](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8367119469/job/22908827843?pr=7874#step:12:1) 🥳!
> 
> image
> 
> image
> 
> Awesome improvement! 🍻 
> 


@maks-rafalko 👏

File: src/AbstractFixer.php
Comment (keradus) [inline src/AbstractFixer.php:1]: Screenshot 2024-03-21 at 11 49 43

should we revert?

Comment (Wirone) [reply src/AbstractFixer.php:1]: Of course 🙂. I did not do it on purpose to prevent merge, as there's still work around `src/Runner/Runner.php` (unless we decide to ignore it, as it's an edge case - someone have to run Infection locally in the first place 😝).

Aaaaaand I don't know if I want to have it merged before #7777 😂.

Comment (keradus) [reply src/AbstractFixer.php:1]: you may also consider to convert to draft, if not ready for merge

Comment (maks-rafalko) [reply src/AbstractFixer.php:1]: > as there's still work around src/Runner/Runner.php (unless we decide to ignore it, as it's an edge case - someone have to run Infection locally in the first place 😝).

could you please explain what is the issue? (trying to understand if there is anything I can help with)

Comment (Wirone) [reply src/AbstractFixer.php:1]: @keradus problematic mutator found and added to ignore list, PR is ready 🙂.

Comment (maks-rafalko) [reply src/AbstractFixer.php:1]: btw, did you get why *source files* were modified? Infection (in normal execution) never modifies original files. What it does - it dumps a new file in `tmpDir` with mutated code.

Comment (Wirone) [reply src/AbstractFixer.php:1]: @maks-rafalko I did not dig deep for now, but I believe it makes dry-run a regular one somehow, which leads to saving modified tokens into fixture files.

BTW. The new option is useful also for running Infection with `--filter`, so I could easily run mutation tests only for `Runner` 🙂.

Comment (maks-rafalko) [reply src/AbstractFixer.php:1]: > BTW. The new option is useful also for running Infection with --filter

right, I documented it as well. I'm glad it helps you :)

**Comment by keradus** [comment]
@Wirone , this failed on master, mind to take a look?
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/actions/runs/8428502226/job/23081111767#step:12:257

**Comment by Wirone** [comment]
Yeah, Infection job should be triggered only for PRs at this point.
